### PR TITLE
AP_Mount: add second instance of mount

### PR DESCRIPTION
**Pull request by rmackay9**
This PR adds support for a 2nd gimbal, and while at first this looks like a big change, in fact, it is only the last 3 commits that are new.  The rest are from this existing PR https://github.com/ArduPilot/ardupilot/pull/21602

Changes included in this PR are:

1. MNT2_xx parameters appear in the parameter list for all vehicles that include AP_Mount
2. DO_GIMBAL_MANAGER_PITCHYAW handling is enhanced to consume the gimbal_id (aka "instance") and pass it to the mount library which already supports multiple gimbals.  This change affects both mission commands and "immediate" commands sent from the ground station.

This is how it behaves:

1. Older commands including MOUNT_CONTROL, MOUNT_CONFIG, SET_ROI have no way to specify the "gimbal id" (aka "instance") so they always affect the "primary" gimbal which is the 1st gimbal.  So for example, if using Mission Planner, clicking on the map and selecting, "Point Camera Here" only affects the first gimbal.  Similarly pushing the "Set Mount" button only affects the first gimbal.
image
2. The MNTx_DEFLT_MODE parameter exists for each instance so if the user always wants the gimbal to point home (for example) this PR is sufficient for them.
3. The DO_GIMBAL_MANAGER_PITCHYAW can be used to point any gimbal instance in a particular direction (in either body-frame or earth-frame).  This command also supports switching to Retracted or Neutral mode through the use of the Flags field
4. RCx_OPTION values exist for the 1st or 2nd gimbal so RC control of either gimbal works

Known limitations:

1. RC control only works for up to 2 gimbals (we would need to add more RCx_OPTION enums to expand this)
2. Multiple Servo and Gremsy gimbals work
3. Two SToRM32 MAVlink gimbals won't work because they "find" using GCS_MAVLINK::find_by_mavtype() which would always find the first instance meaning both drivers would try to control the same gimbal.  We could perhaps resolve this by enhancing the find_by_mavtype() to accept and instance.
4. SToRM32 serial and Alexmos serial gimbals have similar problems to above because they use Serial_Manager::find_serial() to find the serial port for the gimbal.  We could easily resolve this by passing in the instance but then users would struggle to use a mix of gimbal types (e.g. a Servo gimbal and a SToRM32 serial gimbal).  I can easily add the instance in though if others think this would be useful.
5. We don't have a "Mount2 Lock" RC option so the pilot can't switch between earth-frame and body-frame for the 2nd gimbal.  We should add this.

This has been lightly tested in SITL with the following successful results:

- Pilot RC input worked for two gimbals
- "Legacy" commands only affected the first gimbal
- DO_GIMBAL_MANAGER_PITCHYAW could control each gimbal correctly

image

Questions:

- [ ] should we search-and-replace "Mount" to "Gimbal" in the code?  How about MNTx_ params renamed to GIMx_?

### CODE REVIEW

**Comment by IamPete1** [comment]
This may fix https://github.com/ArduPilot/ardupilot/issues/2488 ?

**Comment by rmackay9** [comment]
@IamPete1, yes, I think this resolves that issue.  By the way, the user mentions, "RCx parameters offer mount2 pan/tilt/roll control" and I think these have become SERVOx_FUNCTION enum values as part of the fix to properly separate inputs from outputs.

File: libraries/AP_Mount/AP_Mount.cpp [hunk 1]
```diff
@@ -278,14 +278,21 @@ MAV_RESULT AP_Mount::handle_command_do_gimbal_manager_pitchyaw(const mavlink_com
         return MAV_RESULT_ACCEPTED;
     }
 
-    // To-Do: handle gimbal device id
+    // check gimbal device id.  0 or 1 is 1st gimbal, 2 is 2nd gimbal, etc
+    uint8_t gimbal_instance = 0;
+    if ((packet.param7 > 0) && (packet.param7 <= AP_MOUNT_MAX_INSTANCES)) {
+        gimbal_instance = (uint8_t)packet.param5 - 1;
```
Comment (tridge) [inline libraries/AP_Mount/AP_Mount.cpp:284]: should be param7?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Mount/AP_Mount.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount.cpp]: oops fixed.  I will retest and figure out how I missed this.

**Comment by rmackay9** [comment]
OK, I've fixed this up now and rebased on master.

I slightly changed the calculation of the gimbal_id.  I was wondering if @IamPete1, @peterbarker or @tridge could confirm they're happy with how I did this?  Search for, "check gimbal device id"

File: libraries/AP_Mount/AP_Mount.cpp [hunk 2]
```diff
@@ -278,14 +278,21 @@ MAV_RESULT AP_Mount::handle_command_do_gimbal_manager_pitchyaw(const mavlink_com
         return MAV_RESULT_ACCEPTED;
     }
 
-    // To-Do: handle gimbal device id
+    // check gimbal device id.  0 is primary, 1 is 1st gimbal, 2 is 2nd gimbal, etc
+    if ((packet.param7 < 0) || (packet.param7 > AP_MOUNT_MAX_INSTANCES)) {
+        return MAV_RESULT_FAILED;
+    }
+    const uint8_t gimbal_instance = is_zero(packet.param7) ? 0 : (uint8_t)packet.param7 - 1;
```
Comment (IamPete1) [inline libraries/AP_Mount/AP_Mount.cpp:285]:
```suggestion
    const uint8_t gimbal_instance = is_zero(packet.param7) ? mount->get_primary() : (uint8_t)packet.param7 - 1;
```

Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount.cpp]: fixed.  thanks!

**Pull request merged**